src/Zendesk/Types.hs

```diff
@@ -1,28 +1,113 @@
     , TicketId
     , TicketURL
-    , TicketInfo(..)
-    , TicketTag(..)
+    , TicketInfo (..)
+    , TicketTag (..)
     , parseAgentId
     , parseComments
     , parseTickets
     , renderTicketStatus
+    -- * General configuration
+    , Config (..)
+    , knowledgebasePath
+    , tokenPath
+    , assignToPath
+    , asksZendeskLayer
+    , App
+    , runApp
     ) where
 
 import           Universum
 
 import           Data.Aeson (FromJSON, ToJSON, Value, object, parseJSON, toJSON, withObject, (.:),
                              (.=))
 import           Data.Aeson.Types (Parser)
-
+import           LogAnalysis.Types (Knowledge)
+
+------------------------------------------------------------
+-- Configuration
+------------------------------------------------------------
+
+newtype App a = App (ReaderT Config IO a)
+    deriving ( Applicative
+             , Functor
+             , Monad
+             , MonadReader Config
+             , MonadIO
+             )
+
+runApp :: App a -> Config -> IO a
+runApp (App a) = runReaderT a
+
+-- | The basic configuration.
+data Config = Config
+    { cfgAgentId            :: !Integer
+    -- ^ Zendesk agent id
+    , cfgZendesk            :: !Text
+    -- ^ URL to Zendesk
+    , cfgToken              :: !Text
+    -- ^ Zendesk token
+    , cfgEmail              :: !Text
+    -- ^ Email address of the user the classifier will process on
+    , cfgAssignTo           :: !Integer
+    -- ^ User that will be assigned to after the classifier has done the analysis
+    , cfgKnowledgebase      :: ![Knowledge]
+    -- ^ Knowledgebase
+    , cfgNumOfLogsToAnalyze :: !Int
+    -- ^ Number of files classifier will analyze
+    , cfgIsCommentPublic    :: !Bool
+    -- ^ If the comment is public or not, for a test run we use an internal comment.
+    , cfgZendeskLayer       :: !(ZendeskLayer App)
+    -- ^ The Zendesk API layer. We will ideally move this into a
+    -- separate configuration containing all the layer (yes, there a couple of them).
+    } -- deriving (Eq, Show)
+
+
+-- | Utility function for getting a function of the @ZendeskLayer@.
+-- There are plenty of other alternatives, but this is the simplest
+-- and most direct one.
+asksZendeskLayer :: forall m a. (MonadReader Config m) => (ZendeskLayer App -> a) -> m a
+asksZendeskLayer getter = do
+    Config{..} <- ask
+    pure $ getter cfgZendeskLayer
+
+
+-- TODO(ks): Move these three below to CLI!
+-- | Path to knowledgebase
+knowledgebasePath :: FilePath
+knowledgebasePath = "./knowledgebase/knowledge.csv"
+
+-- | Filepath to token file
+tokenPath :: FilePath
+tokenPath = "./tmp-secrets/token"
+
+-- | Filepath to assign_to file
+assignToPath :: FilePath
+assignToPath = "./tmp-secrets/assign_to"
+
+-- | The Zendesk API interface that we want to expose.
+-- We don't want anything to leak out, so we expose only the most relevant information,
+-- anything relating to how it internaly works should NOT be exposed.
+data ZendeskLayer m = ZendeskLayer
```

Comment:
I can see what it does but I'm not sure how this would enable us to do simple tests. (Again, now enough haskell knowledge!)

Reply:
I guess you can imagine `ZendeskLayer` as an object (if an object is a dictionary of functions) that we pass around.
When we want to use it we use the `Reader` to fetch the function we are interested in.
The enables us to replace the functions anyway we want because they are defined in the `Config` that `Reader` is using.
In other words, if we now run a test and want `listTickets` to return an empty list, we can, because all we need to do is construct a new configuration with that function a simple constant that returns an empty list and we can test the rest of the code.
If this is confusing, the next PR might make more sense since it will deal with tests themselves.